Fix for "Attempted to dequeue multiple cells for the same index path" in table views with Xcode 9

mikelupo opened the pull request:
Fixes #1007

Trying out a fix for issue #1007.
@justinseanmartin @RoyalPineapple please have a look?

Thanks to Larry Gensch (non-github user at this time) who did all of the heavy lifting.

Thanks,
Mike

justinseanmartin commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -247,32 +253,28 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
+
+					@autoreleasepool {
+						// Scroll to the cell and wait for the animation to complete
+						[tableView scrollToRowAtIndexPath:indexPath atScrollPosition:UITableViewScrollPositionNone animated:NO];
```
The intent of the existing logic is to not scroll the tableview unless it is known that an element within that tableview should be made visible. 

Otherwise, we're taking a potentially huge performance penalty to scroll through all elements in table views to try and find an element that happens to be in the view hierarchy after the table view. This will still happen even if the table view is completely offscreen, right?

Have you noticed this when running your full suite of tests? Maybe it isn't so bad due to not animating the tableview scrolling.

mikelupo replied:
Performance hit is minimal.

mikelupo commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -247,32 +253,28 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
+					@autoreleasepool {
+						// Scroll to the cell and wait for the animation to complete
+						[tableView scrollToRowAtIndexPath:indexPath atScrollPosition:UITableViewScrollPositionNone animated:NO];
+						// Note: using KIFRunLoopRunInModeRelativeToAnimationSpeed here may cause tests to stall
```
That comment was in original code.

mikelupo commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -247,32 +253,28 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
+							continue;
+						}
+					}
+					CFRunLoopRunInMode(UIApplicationCurrentRunMode, 0, false);
```
Optimization. Still it ensures that it runs on the correct thread.

mikelupo commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -236,7 +236,13 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
             }];
-            
+
+			NSIndexPath *firstIndexPath = indexPathsForVisibleRows.firstObject;
```
Yeah, I encounter this in just about every github repro. ;) You'd think I'd know by now.

justinseanmartin commented:
Regarding Travis, I think there might be other issues we need to deal with to get those to go green with XC9. 

Last time I tried on XC8 (which we should also have CI coverage of but don't currently), the issues were with autocorrect kicking in and breaking some tests. We could fix them one-off by disabling auto-correct on those fields in the integration test app.

It looks like in your job, the XC9 job is failing a bunch of tests that likely need to be investigated to get that CI configuration added. I'll create an issue for that and track it in the XC9 project. You can remove that from this project for now. 

Ideally, you'd be able to create a test that when run locally against XC9, it shows it failing without the fix and passing with it. That test should also pass in XC6/7/8, so that would be fine. We just can't ensure we don't regress the behavior that is currently broken in XC9 without CI validating it. Having a test is better than nothing though if you understand it well enough.

justinseanmartin commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -248,31 +256,29 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
                         UIAccessibilityElement *element = [cell accessibilityElementMatchingBlock:matchBlock notHidden:NO];
-                        
-                        // Remove the cell from the table view so that it doesn't stick around
-                        [cell removeFromSuperview];
```
This isn't necessary or caused some problem?

mikelupo replied:
It seemed to my iOS developer colleague that it wasn't necessary.

justinseanmartin commented:
I'm still a bit concerned about calling `scroll` on the table view over and over as a solution of finding elements that are offscreen. My two concerns are:

1. Performance impact of general KIF usage that isn't impacted by this bug. Performing a single pass to find if an element is in the view hierarchy needs to be very fast.
2. It exercises the app in ways that a user never would, and could potentially lead to false negatives.

I'll pick up the patch and run our test suite with the change and see what happens. In the meantime, remove the XC9 travis config for now (unless you want to try and get it green before introducing this change), and try and get the existing configurations to go green in Travis.

Having a reproduction case in the integration tests that I can run locally and poke around in would be a huge help for understanding exactly what the issue is that we're trying to tackle.

mikelupo commented:
@justinseanmartin I am working towards a green Travis run in between many other things. One major unresolved sticking point is the photo picker. I know your take is that we should not interact with it and obtain the photo by other programmatic means but I’m not convinced that we should as this makes KIF stand out. 
Oh, lastly, I think the Alert tests are failing too. I’ve not investigated. I may just skip those in iOS 11 but would rather not.

mikelupo commented:
FYI, I've been running with this in my own automation without any failures for over a month now. It seems to perform OK too. I will get the XC9 config out. Let's try to get this merged soon so our community can continue using KIF on iOS 11.

justinseanmartin commented:
That's great! I still have an IOU on running this against our tests. Removing XC9 CI config makes sense as far as this change is concerned.

I'm still wary of the overall approach of scrolling the view to find a matching element, but I'll take a look at it when running our test suite. I'll try to get to that soon.

mikelupo commented:
@justinseanmartin updates made. With your approval, this could get merged as-is.

Thanks,
Mike

mikelupo commented:
@justinseanmartin it looks like our code change broke the picker tests. Looking into it.

edwinlai commented:
Is there any progress on this? I manually pointed KIF to this and it fixes the issue.

mikelupo commented:
I’m not sure it’s ready for the masses but feel free to take the patch and use it. I am going to have to lean on @justinseanmartin for some help/advice here. I’ve not been able to give this patch any love for a while.

justinseanmartin commented:
Yea, this is my bad, I've been swamped at work and haven't had time to test it out to see how it works. I'll try to get to it early next week or see if I can get someone else to help out testing our codebase with this change.

It is good to hear that this works for @edwinlai as well. I still wish that there was a test that could be used to replicate the failure and verify the fix doesn't regress.

soleares commented:
@mikelupo @justinseanmartin This resolved the issue for us. I was able to remove all of the workarounds (1 sec async waits). Sanity restored! Thanks for working on this!

charlyliu commented:
This also solve the tests for us.

sirnacnud commented:
I also am running in to this assert with my tests and the PR from @mikelupo fixed it.

justinseanmartin commented:
Sorry for the delay on getting this merged, been swamped. I'm glad this is working for people, and thanks again @mikelupo for helping address this. I'll see about merging this soon.

It is still unfortunate that we don't have a simple reproduction of the issue. Does someone have a small sample project that shows the issue happening?

lyptt commented:
@justinseanmartin for me at least, simple projects never reproduced the issue. Complex UI layouts that were part of real applications were the only trigger.

kenji21 commented:
Got the issue with an empty tableview (but having a tableHeaderView added within the xib file)

justinseanmartin commented:
@kenji21 - Do you have a sample project that I could look at for the repro?

kenji21 commented:
Very hard to find, but the issue comes this line in my client project:
````
self.headerLabelHeightConstraint.constant = self.tableView.bounds.size.height;
````

If I comment it, or change it to:
````
self.headerLabelHeightConstraint.constant = self.tableView.bounds.size.height / 2.0;
````

then the `Attempted to dequeue multiple cells` disappears...

I tried to reproduce the issue in an empty project, with as few code as possible, but the issue doesn't appears... On my client project, cellForRowAtIndexPath is only called two times (second time making it crash).

So instead of trying to reproduce the issue, I've cloned and checkout the branch fixForIssue1007 of mikelupo/KIF, and I can confirm it fix the crash.

justinseanmartin commented:
This is very helpful, thank you!

Just to make sure I understand, you're using auto-layout to size a label contained within a tableview header, and you don't want that to be any bigger than the tableview itself. Is that right?

If you don't mind helping provide some more information, I'd love to help debug this some more. Would you mind checking out a branch with some extra debugging information added?

https://github.com/justinseanmartin/KIF/tree/jmartin/debugging-table-view-cell

If so, would you mind running the smallest reproduction of the issue that you can and send me the console logs? You can find my email in my github bio.

kenji21 commented:
In fact the data is loaded via a network call. Header label is hidden during loading, and will contains "No data" or will have a "Last update date" and a fixed height of 128pt after data parsed and stored into core data. Labels (in header an in cells) are using custom fonts, but this doesn't helped to reproduce the issue in a separate project (copied view controller xib file to this new project), registered Nib cells to the tableView (as in real project), but issue never appeared in the "smaller" project...
Just mailed the log to you

RtgrV commented:
Works awesome with our project.

mikelupo commented:
@justinseanmartin Finally! This will merge. There were several changes to the files I submitted since last November and I had to really merge this carefully. I removed the .travis.yml commits from this stream and squashed three of my own commits on this branch to make the merge go smoother.
Let's please take aim to get this in. Our users are begging us.

To the rest of the users on this thread, since this commit is in sync with master (as of this writing) do you want to try it out first and give us some feedback before we merge to master?

justinseanmartin commented:
Yep, put a comment in the issue #1007, but I'll reiterate here my apology for dragging my feet on this for so long. Let me take a look quick.

Also, Travis is failing a bunch of tests, so we should get those green before merging.

justinseanmartin commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -236,8 +253,17 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
             }];
-            
+
+			BOOL animationEnabled = [KIFUITestActor testActorAnimationsEnabled];
```
NIT (since you need to push again): Spaces are used consistently throughout the file, please match.

justinseanmartin commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -236,8 +253,17 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
                 if (indexPath) {
                     [indexPathsForVisibleRows addObject:indexPath];
                 }
+                if (firstIndexPath == nil) {
```
NIT: This is effectively the same as the single condition:

```
    if (!firstIndexPath || ([firstIndexPath compare:indexPath] == NSOrderedDescending)) {
        firstIndexPath = indexPath;
    }
```

justinseanmartin commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -225,9 +225,26 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
+            for (UIView *view in [self subviews])
+            {
+                NSString * subViewName = [NSString stringWithFormat:@"%@", [view class]];
+                if ([subViewName containsString:@"UIPicker"] )
```
Should we consider doing this for all table views and not just ones for picker views? (e.g. leave the views scrolled to their original position if they aren't a match). I saw some cases where the UI was scrolled to a position that is different than before.

This isn't necessary, but it would be nice to have consistency with the previous behavior and that of a user who isn't interacting with the UI leaving it unmodified.

mikelupo replied:
I decided to leave this alone. There was a reason we chose to do it this way, and I can't recall the exact thought behind it. We have some mileage on this since last Nov and I don't want to introduce something that breaks the fix.

justinseanmartin commented on Additions/UIView-KIFAdditions.m:
```diff
@@ -249,33 +275,34 @@ - (UIAccessibilityElement *)accessibilityElementMatchingBlock:(BOOL(^)(UIAccessi
+            if (firstIndexPath) {
+                [tableView scrollToRowAtIndexPath:firstIndexPath atScrollPosition:UITableViewScrollPositionTop animated:NO];
+            }
+			//if we're in a picker (scrollView), let's make sure we set the position back to how it was last set.
```
NIT: tabs here again too

mikelupo replied:
Yeah, for some reason, our team uses tabs in all of our projects. Make my community software life hell.

justinseanmartin commented:
Completed the review. Mostly tiny style things, and the one question about scroll position in general.

justinseanmartin commented:
One other thing to call: This does exercise app code differently than before. This change identified an area of the app where we were improperly using cell reuse identifiers in our table views. 

I just wanted to call out that this change might break tests for some people. At least in our app though, it is only doing so in ways where app code is broken and needs fixing.

justinseanmartin commented:
There is another bug that I've known about for a while, but hadn't gotten around to fixing that this change started causing in our app. If a table view cell contains a matching element, and that element is full offscreen but the tableViewCell is partially onscreen, we won't scroll it into view and tapping the element will fail. The changed scrolling behavior in this PR causes that preexisting issue to be tripped in our tests. I took this as an opportunity to fix that issue however. I'll put up a separate PR with that fix.

justinseanmartin commented:
Travis tests are broken and need to be fixed before this can merge. Thanks for following up on my comments.

mikelupo commented:
Specific to this commit or is it a more general thing?




-----Original Message-----
From: Justin Martin <notifications@github.com>
To: kif-framework/KIF <KIF@noreply.github.com>
Cc: mikelupo <mikelupo@aol.com>; Mention <mention@noreply.github.com>
Sent: Thu, Jun 7, 2018 6:33 am
Subject: Re: [kif-framework/KIF] Fix for "Attempted to dequeue multiple cells for the same index path" in table views with Xcode 9 (#1020)



Travis tests are broken and need to be fixed before this can merge. Thanks for following up on my comments.
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

justinseanmartin commented:
It looks like it is specific to this commit. The health status is reported at the top of the README.md for master and is green. It can also be seen here: https://travis-ci.org/kif-framework/KIF/branches

justinseanmartin commented:
I've rekicked the CI builds again to see if they fail in the same way, but it looked like some tests that were broken from the changes made in this PR.

mikelupo commented:
Grrr, custom picker tests need some love now. I won’t be able to tend to this till later tonight eastern time.

justinseanmartin commented:
Interesting, now the picker tests work on all platforms/devices except Xcode 7. Thoughts on what it could be?

mikelupo commented:
Initial investigation points towards it being the iPad2. I am using Xcode 9.2 - iPad2 sim with iOS 8.4. 
100% Reproducible there. 
Not sure what is actually happening though. I probably missed some edge case. I did try to look into it and nothing (even stepping through the code) is jumping out at me. 


It's unfortunate that my fix last night didn't finish this task. I am traveling today and cannot give this the priority it deserves for probably the next few weeks.
I realize that this is a priority for our users. 
Please avail yourself. I am not offended if you (or anyone else) completely rewrites my code.


Mike


-----Original Message-----
From: Justin Martin <notifications@github.com>
To: kif-framework/KIF <KIF@noreply.github.com>
Cc: mikelupo <mikelupo@aol.com>; Mention <mention@noreply.github.com>
Sent: Fri, Jun 8, 2018 1:52 am
Subject: Re: [kif-framework/KIF] Fix for "Attempted to dequeue multiple cells for the same index path" in table views with Xcode 9 (#1020)



Interesting, now the picker tests work on all platforms/devices except Xcode 7. Thoughts on what it could be?
—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub, or mute the thread.

qzim commented:
Any update?

mikelupo commented:
@qzim Yes. I need some help to debug an issue on the iPad 2 picker tests. Ya game? Pull this and run the picker tests that failed in Travis. Unless @justinseanmartin says to move forward as-is I cannot commit this

qzim commented:
Why not drop support for 8.x ?

mikelupo commented:
Might make sense as iOS 11 is out but I can’t make that decision alone.

avalanched commented:
I agree with @qzim, https://data.apteligent.com/ios/ indicates a 0.46% iOS 8 usage
